qemu/qemu.go

```diff
@@ -2667,8 +2763,13 @@ func (config *Config) appendKnobs() {
 
 func (config *Config) appendBios() {
 	if config.Bios != "" {
-		config.qemuParams = append(config.qemuParams, "-bios")
-		config.qemuParams = append(config.qemuParams, config.Bios)
+		if config.Knobs.MemEncrypt {
+			config.qemuParams = append(config.qemuParams, "-drive")
+			config.qemuParams = append(config.qemuParams, fmt.Sprintf("if=pflash,format=raw,unit=0,file=%s,readonly", config.Bios))
```

Comment:
It's not at all clear to me why you need this change for encrypted guests.

Reply:
@dgibson, SEV memory encryption requires`OVMF` firmware instead of `BIOS`.

Reply:
Our `kata-containers/configuration.toml` looks like:

```
# Path to the firmware.
# If you want that qemu uses the default firmware leave this option empty
firmware = "/usr/local/share/qemu/OVMF_CODE.fd.kata"
```

Reply:
Ok, I see why you need to change the guest firmware.  I'm still not clear why that means this complicated `-drive` directive rather than just passing a different file to `-bios`.

Reply:
@dgibson My take away from @sandeepg-gupta email (below) is that using the `-bios` option with an OVMF binary is  discouraged... but should work. However, I tested this patch with `-bios` and qemu fails with the following error:
> ctr: failed to launch qemu: exit status 1, error messages from qemu log: qemu: could not load PC BIOS '/usr/local/share/qemu/OVMF_CODE.fd.kata' 

Which leads me to believe that the `-drive` directive is required.

Reply:
I do have issue with the naming of `config.Knobs.MemEncrypt` as the toggle seem specific to the type of firmware and not `confidential-guest-support=`. Perhaps the right approach is to add `config.OVMF` and rename the function to `appendFirmware()`

Reply:
Ok,  So you definitely need *something* here, but the way it's implemented in the current draft means you have really ugly coupling between the layers: govmm is changing the way it implements `appendBios` depending on the `MemEncrypt` flag, but it's the caller (Kata) actually supplying the different BIOS image here.  So `govmm` is assuming that the setting of the `MemEncrypt` option is tied to the type of BIOS image its being given, which isn't really in its purview to decide.  I think there are 3 options here:

1. Does the `-drive` way of doing it also work for regular BIOSes?  If so, just unconditionally change to using that means instead of `-bios`
2. If not, provide separate methods in `govmm` for adding `-bios` and the `-drive` based firmware control.  Then have Kata use the appropriate one for the type of firmware it wants.
3. Or, move the decision about which firmware to use into govmm: have setting `MemEncrypt` *alone* change the firmware image used (at least by default) without having to call `appendBios` as well.

Reply:
Thanks @dgibson, I fully agree this needs changing. Of your options, I think number 3 makes the most sense. Here's how it works today:
- The kata configuration API does specify bios/ovmf, just simply `firmware =.../OVMF.fd`
- The kata-runtime (in `./virtcontainers/qemu.go`) assumes firmware to be bios: `Bios:        firmwarePath`

For now, the strict requirement of OVMF is specific to using MemEncrypt with the AMD SEV architecture. So the simplest change would be to make the decision happen in `govmm` for all `Type: SevGuest` instances.